### PR TITLE
💻 Cooltt Server

### PR DESCRIPTION
**Pull request by TOTBWF**
## Patch Description
This PR adds the ability to communicate with an external tool (for instance, https://github.com/TOTBWF/coolttviz) when running `cooltt`. To do so, I've added a new option `--server PORT` that allows us to specify that we would like to communicate with an external tool.

Furthermore, I've added a notion of "probe hole" that will send the contents of the hole to the server. It can be triggered with by using `#viz` instead of `?` in place of a hole.

Currently we only support probe holes, but in the future it should be easy to extend the server to add further interaction.

### CODE REVIEW

File: src/frontend/Lex.mll [hunk 1]
```diff
@@ -170,6 +170,8 @@ and real_token = parse
     { HOLE_NAME None }
   | "!"
     { BANG }
+  | "!?"
```
Comment (jonsterling) [inline src/frontend/Lex.mll:173]: I would like to use something different here. Maybe like `#viz` or something? idk.

File: src/frontend/Lex.mll [hunk 2]
```diff
@@ -170,6 +170,8 @@ and real_token = parse
...
+    { PROBE_HOLE }
```
Comment (jonsterling) [inline src/frontend/Lex.mll:174]: Let's call this something different; maybe `VISUALIZE`? not sure.

File: src/frontend/dune [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 (library
  (name Frontend)
- (libraries bantorra cooltt.basis cooltt.cubical cooltt.core menhirLib)
+ (libraries bantorra cooltt.basis cooltt.cubical cooltt.core ezjsonm menhirLib)
```
Comment (jonsterling) [inline src/frontend/dune:9]: @favonia How is the ch'i of this library? ok to include? I can't remember what we've used in the past.

**Review by jonsterling — COMMENTED** [review]
This looks super cool and promising!

File: src/frontend/dune (hunk 3 above)
Comment (favonia) [reply src/frontend/dune:9]: `ezjsonm` is okay. In general, JSON without numbers is okay.

Comment (TOTBWF) [reply src/frontend/dune:9]: This is the same library that was used in `redtt`, which is why I chose it in the first place 🙂

Comment (jonsterling) [reply src/frontend/dune:9]: ok cool! ☺️

File: src/frontend/Lex.mll (hunk 1 above)
Comment (TOTBWF) [reply src/frontend/Lex.mll:173]: Sounds good! In prep for this I've bitten the bullet and implemented #281.

**Comment by TOTBWF** [comment]
The change from `!?` to `#viz` is blocked on #281.

**Comment by favonia** [comment]
Also the emoji for this PR could be improved.

**Review by favonia — APPROVED** [review]

**Review by jonsterling — APPROVED** [review]

**Comment by ralsei** [comment]
This PR is now hijacked for the progression of ~~cubical Scratch~~ coolttviz as an editor, and due to the structure of that editor, is now blocked on #386.

**Comment by ralsei** [comment]
The plan is to merge this into main (because it's a massive pain to constantly have to update it, as evidenced by the last commit) and then after that we can work on serialization and other stuff for the visualizer in another branch.

**Review by jonsterling — APPROVED** [review]
Godspeed!

**Pull request merged**